tests/0143-exponential_backoff.c

```diff
@@ -0,0 +1,571 @@
+
+        producer = test_create_handle(RD_KAFKA_PRODUCER, conf);
+        rkt      = test_create_producer_topic(producer, topic, NULL);
+
+        rd_kafka_mock_push_request_errors(
+            mcluster, RD_KAFKAP_Produce, 7,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS,
+            RD_KAFKA_RESP_ERR_COORDINATOR_LOAD_IN_PROGRESS);
+
+        test_produce_msgs(producer, rkt, 0, RD_KAFKA_PARTITION_UA, 0, 1,
+                          "hello", 5);
+        rd_sleep(3);
+
+        err = helper_exponential_backoff(mcluster, RD_KAFKAP_Produce);
+        if (err == -3)
+                TEST_FAIL("Time difference is not respected!");
+
+        rd_kafka_topic_destroy(rkt);
+        rd_kafka_destroy(producer);
+        rd_kafka_mock_clear_requests(mcluster);
+        SUB_TEST_PASS();
+}
+
+/**
+ * Helper function for find coordinator trigger with the given request_type, the
+ * find coordinator should be triggered after failing of request_type.
+ */
+int helper_find_coordinator_trigger(rd_kafka_mock_cluster_t *mcluster,
+                                    int32_t request_type) {
+        rd_kafka_mock_request_t **requests = NULL;
+        size_t request_cnt                 = 0;
+        int32_t num_request                = 0;
+        int16_t error_code                 = 0;
+        size_t i;
+        requests = rd_kafka_mock_get_requests(mcluster, &request_cnt);
+        for (i = 0; i < request_cnt; i++) {
+                TEST_SAY("Broker Id : %d API Key : %d Timestamp : %" PRId64
+                         "\n",
+                         rd_kafka_mock_request_id(requests[i]),
+                         rd_kafka_mock_request_api_key(requests[i]),
+                         rd_kafka_mock_request_timestamp(requests[i]));
+                if (num_request == 0) {
+                        if (rd_kafka_mock_request_api_key(requests[i]) ==
+                            request_type) {
+                                num_request++;
+                        }
+                } else if (num_request == 1) {
+                        if (rd_kafka_mock_request_api_key(requests[i]) ==
+                            RD_KAFKAP_FindCoordinator) {
+                                TEST_SAY(
+                                    "FindCoordinator request made after "
+                                    "failing Heartbeat with NOT_COORDINATOR "
+                                    "error.\n");
+                                break;
+                        } else if (rd_kafka_mock_request_api_key(requests[i]) ==
+                                   request_type) {
+                                num_request++;
+                                error_code = -1;
```

Comment:
As in the above function, don't use error code, just TEST_ASSERT and TEST_FAIL in these functions themselves.